src/providers/signatureHelpProvider.ts

```diff
@@ -0,0 +1,155 @@
+import * as vscode from 'vscode';
+import { HoverRequest, LanguageClient } from 'vscode-languageclient';
+
+export class SignatureHelpProvider implements vscode.SignatureHelpProvider {
+  private languageClient: LanguageClient;
+  private previousFunctionPosition?: vscode.Position;
+
+  constructor(lc: LanguageClient) {
+    this.languageClient = lc;
+  }
+
+  public provideSignatureHelp(
+    document: vscode.TextDocument,
+    position: vscode.Position,
+    token: vscode.CancellationToken,
+    context: vscode.SignatureHelpContext,
+  ): vscode.ProviderResult<vscode.SignatureHelp> {
+    // the current signature help provider uses the hover information from RLS
+    // and it only has a string representation of the function signature.
+    // This check makes sure we can easily show the tooltip for multiple parameters, separated by `,`
+    if (context.triggerCharacter === '(') {
+      this.previousFunctionPosition = position;
+      return this.provideHover(
+        this.languageClient,
+        document,
+        position,
+        token,
+      ).then(hover => {
+        return this.hoverToSignatureHelp(hover);
+      });
+    } else if (context.triggerCharacter === ',') {
+      if (this.previousFunctionPosition) {
+        return this.provideHover(
+          this.languageClient,
+          document,
+          this.previousFunctionPosition,
+          token,
+        ).then(hover => {
+          return this.hoverToSignatureHelp(hover);
+        });
+      } else {
+        return null;
+      }
+    } else {
+      if (context.isRetrigger === false) {
+        this.previousFunctionPosition = undefined;
+      }
+      return null;
+    }
+  }
+
+  private provideHover(
+    lc: LanguageClient,
+    document: vscode.TextDocument,
+    position: vscode.Position,
+    token: vscode.CancellationToken,
+  ): Promise<vscode.Hover> {
+    return new Promise((resolve, reject) => {
+      lc.sendRequest(
```

Comment:
I know that VSCode mostly uses a `Thenable<T>` interface but does it support `await`ing?
If so, I think it'd be good to
```ts
const response = await lc.sendRequest(...);
return lc.protocol2CodeConverter.asHover(response);
```
(assuming it'd throw an error correctly)

Reply:
If we were to `await` here, the containing function, `provideSignatureHelp`, would need to change its signature and become `async` -- which I think means it would no longer correctly implement the `vscode.SignatureHelpProvider` interface.

Reply:
Argh, that might be the case, you’re right. It’s unfortunate the API is settled on the lowest promise common denominator at the time but there’s probably nothing we can do at the time cleanly